rts/motoko-rts/src/region.rs

```diff
@@ -0,0 +1,854 @@
+        loop {
+            let (s_, _, b_len) = r.relative_into_absolute_info(offset + i);
+            s = s_;
+            if i + b_len > dst.len() as u64 {
+                // case: last (generally partial) block.
+                if dst.len() as u64 > i {
+                    read(
+                        s,
+                        core::slice::from_raw_parts_mut(d, dst.len() - i as usize),
+                    );
+                }
+                break;
+            } else {
+                // case: internal (full) block.
+                read(
+                    s,
+                    core::slice::from_raw_parts_mut(d, BLOCK_IN_BYTES as usize),
+                );
+                d = d.offset(BLOCK_IN_BYTES as isize);
+                i += BLOCK_IN_BYTES;
+            }
+        }
+    }
+}
+
+pub(crate) unsafe fn region_store<M: Memory>(_mem: &mut M, r: Value, offset: u64, src: &[u8]) {
+    use crate::ic0_stable::nicer::write;
+    use meta_data::size::BLOCK_IN_BYTES;
+
+    let r = RegionObject::from_value(&r);
+    if src.len() > 1 {
+        r.check_relative_into_absolute_offset(offset + src.len() as u64 - 1);
+    } else {
+        r.check_relative_into_absolute_offset(offset);
+    }
+    let (b1_off, b1, b1_len, b2) = r.relative_into_absolute_span(offset, src.len() as u64);
+    if b1 == b2 {
+        write(b1_off, src);
+    } else {
+        // Case: Staged reads, one per block that holds requested data.
+
+        let mut i = 0; // invariant: i = # of bytes stored.
+        let mut s = src.as_ptr(); // source for bytes.
+        let mut d = b1_off; // dest of bytes, as absolute index.o
+
+        // do initial write (a special case, generally not full block length).
+        write(d, core::slice::from_raw_parts(s, b1_len as usize));
+
+        // Advance input and output positions (i, s and d respectively).
+        i += b1_len;
+        s = s.offset(b1_len as isize);
+
+        // Do rest of block-sized reads.
+        // (invariant: they always occur at the start of a block).
+        loop {
+            let (d_, _, b_len) = r.relative_into_absolute_info(offset + i);
+            d = d_;
+            if i + b_len > src.len() as u64 {
+                // case: last (generally partial) block.
+                if src.len() as u64 > i {
+                    write(d, core::slice::from_raw_parts(s, src.len() - i as usize));
+                }
+                break;
+            } else {
+                // case: internal (full) block.
+                write(d, core::slice::from_raw_parts(s, BLOCK_IN_BYTES as usize));
+                s = s.offset(BLOCK_IN_BYTES as isize);
+                i += BLOCK_IN_BYTES;
+            }
+        }
+    }
+}
+
+// -- Region load operations.
+
+#[ic_mem_fn]
+pub unsafe fn region_load_word8<M: Memory>(mem: &mut M, r: Value, offset: u64) -> u32 {
+    let mut byte: [u8; 1] = [0];
+    region_load(mem, r, offset, &mut byte);
+    byte[0] as u32
```

Comment:
Detail (unimportant, not sure whether it is worth it): Possible symmetry to the other load functions:
```suggestion
    let mut bytes: [u8; 1] = [0];
    region_load(mem, r, offset, &mut bytes);
    core::primitive::u8::from_le_bytes(bytes).into()
```